File: internal/core/ec2manager.go
package core

import (
	"context"
	"fmt"
	"sort"
	"sync"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/ec2"
	"github.com/aws/aws-sdk-go-v2/service/ec2/types"
)

var RegionNames = map[string]string{
	"us-east-1":      "N. Virginia",
	"us-east-2":      "Ohio",
	"us-west-1":      "N. California",
	"us-west-2":      "Oregon",
	"ap-south-1":     "Mumbai",
	"ap-northeast-1": "Tokyo",
	"ap-northeast-2": "Seoul",
	"ap-northeast-3": "Osaka",
	"ap-southeast-1": "Singapore",
	"ap-southeast-2": "Sydney",
	"ca-central-1":   "Canada",
	"eu-central-1":   "Frankfurt",
	"eu-west-1":      "Ireland",
	"eu-west-2":      "London",
	"eu-west-3":      "Paris",
	"eu-north-1":     "Stockholm",
}

type Instance struct {
	ID                    string
	Name                  string
	Status                string
	InstanceType          string
	Region                string
	TerminationProtection bool
	StopProtection        bool
}

type EC2Manager struct {
	accessKey string
	secretKey string
	clients   map[string]*ec2.Client
	mu        sync.RWMutex
}

func NewEC2Manager(accessKey, secretKey string) *EC2Manager {
	return &EC2Manager{
		accessKey: accessKey,
		secretKey: secretKey,
		clients:   make(map[string]*ec2.Client),
	}
}

func (m *EC2Manager) getClient(region string) (*ec2.Client, error) {
	m.mu.RLock()
	if client, ok := m.clients[region]; ok {
		m.mu.RUnlock()
		return client, nil
	}
	m.mu.RUnlock()

	cfg, err := config.LoadDefaultConfig(context.TODO(),
		config.WithRegion(region),
		config.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(
			m.accessKey,
			m.secretKey,
			"",
		)),
	)
	if err != nil {
		return nil, err
	}

	client := ec2.NewFromConfig(cfg)
	
	m.mu.Lock()
	m.clients[region] = client
	m.mu.Unlock()

	return client, nil
}

func (m *EC2Manager) GetAvailableRegions() ([]string, error) {
	client, err := m.getClient("us-east-1")
	if err != nil {
		return nil, err
	}

	result, err := client.DescribeRegions(context.TODO(), &ec2.DescribeRegionsInput{})
	if err != nil {
		return nil, err
	}

	regions := make([]string, 0, len(result.Regions))
	for _, r := range result.Regions {
		regions = append(regions, *r.RegionName)
	}

	return regions, nil
}

func (m *EC2Manager) GetInstancesInRegion(region string) ([]Instance, error) {
	client, err := m.getClient(region)
	if err != nil {
		return nil, err
	}

	result, err := client.DescribeInstances(context.TODO(), &ec2.DescribeInstancesInput{})
	if err != nil {
		return nil, err
	}

	var instances []Instance
	for _, reservation := range result.Reservations {
		for _, inst := range reservation.Instances {
			instance := m.parseInstance(inst, region)
			m.fetchProtectionInfo(client, &instance)
			instances = append(instances, instance)
		}
	}

	return instances, nil
}

func (m *EC2Manager) GetAllInstances() ([]Instance, error) {
	regions, err := m.GetAvailableRegions()
	if err != nil {
		return nil, err
	}

	var (
		instances []Instance
		mu        sync.Mutex
		wg        sync.WaitGroup
	)

	for _, region := range regions {
		wg.Add(1)
		go func(r string) {
			defer wg.Done()
			
			insts, err := m.GetInstancesInRegion(r)
			if err != nil {
				return
			}

			mu.Lock()
			instances = append(instances, insts...)
			mu.Unlock()
		}(region)
	}

	wg.Wait()

	sort.Slice(instances, func(i, j int) bool {
		if instances[i].Name == "" && instances[j].Name != "" {
			return false
		}
		if instances[i].Name != "" && instances[j].Name == "" {
			return true
		}
		if instances[i].Name != instances[j].Name {
			return instances[i].Name < instances[j].Name
		}
		return instances[i].ID < instances[j].ID
	})

	return instances, nil
}

func (m *EC2Manager) StartInstances(instanceIDs []string, region string) error {
	client, err := m.getClient(region)
	if err != nil {
		return err
	}

	_, err = client.StartInstances(context.TODO(), &ec2.StartInstancesInput{
		InstanceIds: instanceIDs,
	})
	return err
}

func (m *EC2Manager) StopInstances(instanceIDs []string, region string) error {
	client, err := m.getClient(region)
	if err != nil {
		return err
	}

	_, err = client.StopInstances(context.TODO(), &ec2.StopInstancesInput{
		InstanceIds: instanceIDs,
	})
	return err
}

func (m *EC2Manager) TerminateInstances(instanceIDs []string, region string) error {
	client, err := m.getClient(region)
	if err != nil {
		return err
	}

	_, err = client.TerminateInstances(context.TODO(), &ec2.TerminateInstancesInput{
		InstanceIds: instanceIDs,
	})
	return err
}

func (m *EC2Manager) SetTerminationProtection(instanceID, region string, enabled bool) error {
	client, err := m.getClient(region)
	if err != nil {
		return err
	}

	_, err = client.ModifyInstanceAttribute(context.TODO(), &ec2.ModifyInstanceAttributeInput{
		InstanceId: aws.String(instanceID),
		DisableApiTermination: &types.AttributeBooleanValue{
			Value: aws.Bool(enabled),
		},
	})
	return err
}

func (m *EC2Manager) SetStopProtection(instanceID, region string, enabled bool) error {
	client, err := m.getClient(region)
	if err != nil {
		return err
	}

	_, err = client.ModifyInstanceAttribute(context.TODO(), &ec2.ModifyInstanceAttributeInput{
		InstanceId: aws.String(instanceID),
		DisableApiStop: &types.AttributeBooleanValue{
			Value: aws.Bool(enabled),
		},
	})
	return err
}

func (m *EC2Manager) parseInstance(inst types.Instance, region string) Instance {
	name := ""
	for _, tag := range inst.Tags {
		if *tag.Key == "Name" {
			name = *tag.Value
			break
		}
	}

	return Instance{
		ID:           *inst.InstanceId,
		Name:         name,
		Status:       string(inst.State.Name),
		InstanceType: string(inst.InstanceType),
		Region:       region,
	}
}

func (m *EC2Manager) fetchProtectionInfo(client *ec2.Client, inst *Instance) {
	termResult, _ := client.DescribeInstanceAttribute(context.TODO(), &ec2.DescribeInstanceAttributeInput{
		InstanceId: aws.String(inst.ID),
		Attribute:  types.InstanceAttributeNameDisableApiTermination,
	})
	if termResult != nil && termResult.DisableApiTermination != nil {
		inst.TerminationProtection = *termResult.DisableApiTermination.Value
	}

	stopResult, _ := client.DescribeInstanceAttribute(context.TODO(), &ec2.DescribeInstanceAttributeInput{
		InstanceId: aws.String(inst.ID),
		Attribute:  types.InstanceAttributeNameDisableApiStop,
	})
	if stopResult != nil && stopResult.DisableApiStop != nil {
		inst.StopProtection = *stopResult.DisableApiStop.Value
	}
}

func GetRegionDisplayName(region string) string {
	if name, ok := RegionNames[region]; ok {
		return fmt.Sprintf("%s (%s)", region, name)
	}
	return region
}

File: src/core/ec2_manager.py
"""
AWS EC2 관리 핵심 로직
UI와 독립적으로 작동 → 테스트 가능
"""
import boto3
from typing import List, Dict, Optional
from concurrent.futures import ThreadPoolExecutor, as_completed


class EC2Instance:
    """EC2 인스턴스 정보"""
    def __init__(self, instance_id: str, name: str, status: str, 
                 instance_type: str, region: str,
                 termination_protection: bool = False,
                 stop_protection: bool = False):
        self.id = instance_id
        self.name = name
        self.status = status
        self.type = instance_type
        self.region = region
        self.termination_protection = termination_protection
        self.stop_protection = stop_protection
    
    def to_dict(self) -> Dict:
        """딕셔너리로 변환"""
        return {
            'id': self.id,
            'name': self.name,
            'status': self.status,
            'type': self.type,
            'region': self.region,
            'termination_protection': self.termination_protection,
            'stop_protection': self.stop_protection
        }


class EC2Manager:
    """AWS EC2 관리 클래스 (비즈니스 로직만)"""
    
    REGION_NAMES = {
        'us-east-1': 'N. Virginia',
        'us-east-2': 'Ohio',
        'us-west-1': 'N. California',
        'us-west-2': 'Oregon',
        'ap-south-1': 'Mumbai',
        'ap-northeast-1': 'Tokyo',
        'ap-northeast-2': 'Seoul',
        'ap-northeast-3': 'Osaka',
        'ap-southeast-1': 'Singapore',
        'ap-southeast-2': 'Sydney',
        'ap-southeast-3': 'Jakarta',
        'ap-southeast-4': 'Melbourne',
        'ap-east-1': 'Hong Kong',
        'ca-central-1': 'Canada',
        'ca-west-1': 'Calgary',
        'eu-central-1': 'Frankfurt',
        'eu-central-2': 'Zurich',
        'eu-west-1': 'Ireland',
        'eu-west-2': 'London',
        'eu-west-3': 'Paris',
        'eu-north-1': 'Stockholm',
        'eu-south-1': 'Milan',
        'eu-south-2': 'Spain',
        'me-south-1': 'Bahrain',
        'me-central-1': 'UAE',
        'sa-east-1': 'São Paulo',
        'af-south-1': 'Cape Town',
        'il-central-1': 'Tel Aviv'
    }
    
    def __init__(self, access_key: str, secret_key: str):
        self.access_key = access_key
        self.secret_key = secret_key
        self._regions_cache = None
    
    def create_client(self, region: str):
        """EC2 클라이언트 생성"""
        return boto3.client(
            'ec2',
            aws_access_key_id=self.access_key,
            aws_secret_access_key=self.secret_key,
            region_name=region
        )
    
    def get_available_regions(self) -> List[str]:
        """사용 가능한 AWS 리전 목록"""
        if self._regions_cache:
            return self._regions_cache
        
        ec2 = self.create_client('us-east-1')
        response = ec2.describe_regions()
        self._regions_cache = [r['RegionName'] for r in response['Regions']]
        return self._regions_cache
    
    def get_instances_in_region(self, region: str) -> List[EC2Instance]:
        """특정 리전의 인스턴스 조회"""
        ec2 = self.create_client(region)
        response = ec2.describe_instances()
        
        instances = []
        for reservation in response['Reservations']:
            for instance in reservation['Instances']:
                instances.append(self._parse_instance(instance, region))
        
        # Protection 정보 병렬 조회
        self._fetch_protection_info(instances)
        
        return instances
    
    def get_all_instances(self, regions: Optional[List[str]] = None) -> List[EC2Instance]:
        """모든 리전의 인스턴스 조회 (병렬)"""
        if not regions:
            regions = self.get_available_regions()
        
        all_instances = []
        
        with ThreadPoolExecutor(max_workers=10) as executor:
            future_to_region = {
                executor.submit(self.get_instances_in_region, region): region 
                for region in regions
            }
            
            for future in as_completed(future_to_region):
                try:
                    instances = future.result()
                    all_instances.extend(instances)
                except Exception:
                    # 특정 리전 실패해도 계속 진행
                    pass
        
        return all_instances
    
    def start_instances(self, instance_ids: List[str], region: str):
        """인스턴스 시작"""
        ec2 = self.create_client(region)
        ec2.start_instances(InstanceIds=instance_ids)
    
    def stop_instances(self, instance_ids: List[str], region: str):
        """인스턴스 중지"""
        ec2 = self.create_client(region)
        ec2.stop_instances(InstanceIds=instance_ids)
    
    def terminate_instances(self, instance_ids: List[str], region: str):
        """인스턴스 삭제"""
        ec2 = self.create_client(region)
        ec2.terminate_instances(InstanceIds=instance_ids)
    
    def set_termination_protection(self, instance_id: str, region: str, enabled: bool):
        """Termination Protection 설정"""
        ec2 = self.create_client(region)
        ec2.modify_instance_attribute(
            InstanceId=instance_id,
            DisableApiTermination={'Value': enabled}
        )
    
    def set_stop_protection(self, instance_id: str, region: str, enabled: bool):
        """Stop Protection 설정"""
        ec2 = self.create_client(region)
        ec2.modify_instance_attribute(
            InstanceId=instance_id,
            DisableApiStop={'Value': enabled}
        )
    
    def _parse_instance(self, instance: Dict, region: str) -> EC2Instance:
        """AWS 응답을 EC2Instance 객체로 변환"""
        name = ''
        for tag in instance.get('Tags', []):
            if tag['Key'] == 'Name':
                name = tag['Value']
                break
        
        return EC2Instance(
            instance_id=instance['InstanceId'],
            name=name,
            status=instance['State']['Name'],
            instance_type=instance['InstanceType'],
            region=region
        )
    
    def _fetch_protection_info(self, instances: List[EC2Instance]):
        """Protection 정보 병렬 조회"""
        def fetch_protection(inst: EC2Instance):
            try:
                ec2 = self.create_client(inst.region)
                
                # Termination protection
                resp = ec2.describe_instance_attribute(
                    InstanceId=inst.id, 
                    Attribute='disableApiTermination'
                )
                inst.termination_protection = resp.get('DisableApiTermination', {}).get('Value', False)
                
                # Stop protection
                resp = ec2.describe_instance_attribute(
                    InstanceId=inst.id,
                    Attribute='disableApiStop'
                )
                inst.stop_protection = resp.get('DisableApiStop', {}).get('Value', False)
            except Exception:
                pass
        
        with ThreadPoolExecutor(max_workers=10) as executor:
            executor.map(fetch_protection, instances)
    
    @staticmethod
    def get_region_display_name(region_code: str) -> str:
        """리전 코드 → 표시명"""
        if region_code in EC2Manager.REGION_NAMES:
            return f"{region_code} ({EC2Manager.REGION_NAMES[region_code]})"
        return region_code
    
    @staticmethod
    def extract_region_code(region_display: str) -> str:
        """표시명 → 리전 코드"""
        return region_display.split(' (')[0] if ' (' in region_display else region_display
